File: Behavioral/Lab14/C#/Builders/EstateBuilder.cs
using System;
using System.Collections.Generic;
using System.Text;
using C_.Interfaces;
using C_.Entities;
namespace C_.Builders
{
    public abstract class EstateBuilder
    {
        protected RealEstate _estate;
        public abstract void Reset();
        public RealEstate GetEstate(){
            var clone = (RealEstate)_estate.Clone();
            Reset();
            return clone;
        }

        public EstateBuilder(){
            Reset();
        }

        public virtual EstateBuilder MakeSellable(){
            _estate.IsSellable = true;
            return this;
        }
        public virtual EstateBuilder MakeBuyable(){
            _estate.IsBuyable = true;
            return this;
        }
        public virtual EstateBuilder MakeRentable(){
            _estate.IsRentable = true;
            return this;
        }
        public virtual EstateBuilder AddRealtor(RealtorAbstract realtor){
            _estate.Realtor = realtor;
            return this;
        }
        public virtual EstateBuilder SetStatus(string status){
            _estate.Status = status;
            return this;
        }
        public virtual EstateBuilder SetPrice(double price){
            _estate.PriceInUSD = price;
            return this;
        }
        public virtual EstateBuilder SetArea(double area){
            _estate.Area = area;
            return this;
        }
        public virtual EstateBuilder SetDescription(string description){
            _estate.Description = description;
            return this;
        }
        public virtual EstateBuilder SetAddress(string address){
            _estate.Address = address;
            return this;
        }
        ///Bools:
        public virtual EstateBuilder SetFurnished(){
            _estate.Furnished = true;
            return this;
        }
        public virtual EstateBuilder SetParking(){
            _estate.Parking = true;
            return this;
        }
        public virtual EstateBuilder SetBalcony(){
            _estate.Balcony = true;
            return this;
        }
        public virtual EstateBuilder SetElevator(){
            _estate.Elevator = true;
            return this;
        }
        public virtual EstateBuilder SetAirConditioning(){
            _estate.AirConditioning = true;
            return this;
        }
        public virtual EstateBuilder SetUtilitiesIncluded(){
            _estate.IsUtilitiesIncluded = true;
            return this;
        }
    }
}

File: Behavioral/Lab19/Typescript/src/Builders/estate_builder.ts
import { RealEstate } from '../Entities/real_estate';
import { Realtor } from '../Entities/realtor';

export abstract class EstateBuilder {
    protected _estate: RealEstate;

    public abstract reset(): void;

    public getEstate(): RealEstate {
        const clone = this._estate.clone();
        this.reset();
        return clone;
    }

    constructor() {
        this.reset();
    }

    public makeSellable(): EstateBuilder {
        this._estate.isSellable = true;
        return this;
    }

    public makeBuyable(): EstateBuilder {
        this._estate.isBuyable = true;
        return this;
    }

    public makeRentable(): EstateBuilder {
        this._estate.isRentable = true;
        return this;
    }

    public addRealtor(realtor: Realtor): EstateBuilder {
        this._estate.realtor = realtor;
        return this;
    }

    public setStatus(status: string): EstateBuilder {
        this._estate.status = status;
        return this;
    }

    public setPrice(price: number): EstateBuilder {
        this._estate.priceInUSD = price;
        return this;
    }

    public setArea(area: number): EstateBuilder {
        this._estate.area = area;
        return this;
    }

    public setDescription(description: string): EstateBuilder {
        this._estate.description = description;
        return this;
    }

    public setAddress(address: string): EstateBuilder {
        this._estate.address = address;
        return this;
    }

    public setFurnished(): EstateBuilder {
        this._estate.furnished = true;
        return this;
    }

    public setParking(): EstateBuilder {
        this._estate.parking = true;
        return this;
    }

    public setBalcony(): EstateBuilder {
        this._estate.balcony = true;
        return this;
    }

    public setElevator(): EstateBuilder {
        this._estate.elevator = true;
        return this;
    }

    public setAirConditioning(): EstateBuilder {
        this._estate.airConditioning = true;
        return this;
    }

    public setUtilitiesIncluded(): EstateBuilder {
        this._estate.isUtilitiesIncluded = true;
        return this;
    }
}